Improve handling of non-BMP attribute names

lucaswerkmeister opened the pull request:
This incorporates two changes:

1. `Naming.capitalize` now correctly handles characters outside the Basic Multilingual Plane, that is, characters that don't fit within a single Java `char`. Theoretically, this breaks binary compatibility: if the name of an attribute begins with a character outside the BMP that has an uppercase variant (for example, a lowercase Deseret character), then the getter is now called `getX` (where `X` is the uppercase variant), whereas previously it was called `getx` (`x` being the lowercase variant). However, I find it very unlikely that anyone would actually have done that.
2. `AbstractModelLoader.isStartOfJavaBeanPropertyName` now uses codepoints instead of characters as well; in addition, its logic has been slightly changed: instead of checking whether the character after the "get" is uppercased, it now checks whether that character is equal to its uppercase variant. This makes a difference if the character does not have an uppercase variant, i. e., it is its own uppercase version. This part fixes ceylon/ceylon.language#634.

I’m getting some test failures here; I’m mostly confident that they’re unrelated to this change, but someone else should definitely still check that.

tombentley commented:
I think we should merge this, but:

1. Could we have a PR for the tests in the gist linked-to from ceylon/ceylon.language#634
2. Do you know whether the js compiler already handles this stuff correctly, or whether we'll need to change `ceylon-js` in a similar fashion to this?

lucaswerkmeister commented:
@tombentley I added a test (and rebased onto master). I’m getting some test failures here (including one that looks metamodel-related, `bugJ505`), but it looks like I’m getting them all on `master` too, so I don’t think they’re related.

tombentley commented:
Thanks @lucaswerkmeister!

tombentley commented:
@lucaswerkmeister having merged these patches I'm finding that testBugCL634 is failing. More fool me for not actually testing before merging, but I'm be grateful if you could help fix the problem.

lucaswerkmeister commented:
Okay, I’ll look into it when I get home around 17:00 CET

lucaswerkmeister commented:
@tombentley Sorry, I must’ve missed that. It’s not in the `ant test` output, but it appears clearly in the HTML test reports. I’ll look into it.

lucaswerkmeister commented:
1. Part of the failure was just stupidity on my part: I hadn’t adjusted the entire test for the additional member. I fixed that in 335846a.
2. It appears that when a class is `@noanno`’ed, `declaredMemberDeclarations()` yields the `getType()` function and the `$TypeDescriptor$` value as well. Not sure if that’s my fault.
3. The Deseret member is reported as the members `value tmp::BugCL634.𐐀, value tmp::BugCL634.𐐨`. Not sure why it’s duplicate, but that *is* almost certainly my fault. Will look into it.

lucaswerkmeister commented:
Okay, what’s happening in 3 is that the `Declaration` `JavaBeanValue[BugCL634.А:Object]` has the *uppercase* `𐐀` as the name. Therefore, it can’t be matched with the already loaded `𐐨` value, so it’s added again (duplicate). I need to find out where that name fails to be lowercased again.

lucaswerkmeister commented:
Found it, [`CodegenUtil.getJavaBeanName`](https://github.com/ceylon/ceylon-compiler/blob/0e485de8729fa0f9eda114557e29b88ea0f41895/src/com/redhat/ceylon/compiler/java/codegen/CodegenUtil.java#L539-L555) is working with `char`s too.

lucaswerkmeister commented:
3 fixed in 1ed07a9. 2 remains: the `@noanno`’ed class sees some hidden methods. I’ll try to find out whether that’s my fault or not :)

quintesse commented:
Not sure if I'm talking nonsense but it could be because the `@noanno` annotation is also preventing the `@Ignore` annotations from being generated thereby changing the class' meta model.

lucaswerkmeister commented:
Hm, you’re right, no `@Ignore` in the `.src`. I guess this is expected, then.

quintesse commented:
In any case classes with `@noanno` should never actually be executed, they are only meant for code comparisons. All kinds of essential meta data will be missing from the class.

lucaswerkmeister commented:
You’re completely right, I’m didn’t think about that at all. Thanks.

@tombentley test works for me now.

> More fool me for not actually testing before merging

I hope when we get OBR we’ll also have a build server that automatically tests and [tags](https://developer.github.com/v3/repos/statuses/) commits as failing or succeeding… that would be awesome
